### PR TITLE
[WIP] Tagged pointers

### PR DESCRIPTION
**Pull request by vstinner**
See added TAGGED_POINTERS.rst for more information.

### CODE REVIEW

**Comment by vstinner** [comment]
Here is an experimental implementation of tagged pointers for CPython. So far, I implemented None, True, False and small integers. I may be interesting to implement more types like float or even short latin1 strings (latin1 single character singletons?).

The PR is large because of the first changes which add Py_IS_NONE(), Py_IS_TRUE() and Py_IS_FALSE() macros.

I didn't run any benchmark yet.

cc @nascheme @pitrou @methane @serhiy-storchaka @pablogsal

File: Include/object.h [hunk 1]
```diff
@@ -122,43 +122,95 @@ typedef struct {
 #define _PyVarObject_CAST_CONST(op) ((const PyVarObject*)(op))
 
 
+// On x86-64, memory allocations must be aligned on 8 bytes (2**3)
```
Comment (pitrou) [inline Include/object.h:125]: Or even 16? I don't remember.

File: Include/object.h [hunk 2]
```diff
@@ -122,43 +122,95 @@ typedef struct {
...
 #define Py_REFCNT(ob) _Py_REFCNT(_PyObject_CAST_CONST(ob))
 
 
 static inline Py_ssize_t _Py_SIZE(const PyVarObject *ob) {
-    return ob->ob_size;
+    if (!_Py_TAGPTR_IS_TAGGED((PyObject*)ob)) {
+        return ob->ob_size;
+    }
+    else {
+        PyVarObject *unboxed = (PyVarObject *)_Py_TAGPTR_UNBOX((PyObject*)ob);
```
Comment (pitrou) [inline Include/object.h:181]: I suppose this is temporary? Most tagged values will have size 0 (or perhaps 1 if you also encode single-char strings), there's no need to access the "actual" object.

File: Objects/longobject.c [hunk 3]
```diff
@@ -26,10 +26,21 @@ _Py_IDENTIFIER(little);
 _Py_IDENTIFIER(big);
 
 /* convert a PyLong of size 1, 0 or -1 to an sdigit */
-#define MEDIUM_VALUE(x) (assert(-1 <= Py_SIZE(x) && Py_SIZE(x) <= 1),   \
-         Py_SIZE(x) < 0 ? -(sdigit)(x)->ob_digit[0] :   \
-             (Py_SIZE(x) == 0 ? (sdigit)0 :                             \
-              (sdigit)(x)->ob_digit[0]))
+static inline sdigit MEDIUM_VALUE(PyLongObject *x)
+{
+    x = (PyLongObject *)_Py_TAGPTR_UNBOX((PyObject*)x);
```
Comment (pitrou) [inline Objects/longobject.c:31]: Why isn't `x` a tagged integer here? You could comfortably embed a `sdigit` in a tagged pointer.

File: Include/object.h (hunk 2 above)
Comment (vstinner) [reply Include/object.h:181]: I first tried to be correct: except test_gdb, the whole test suite pass.

Comment (vstinner) [reply Include/object.h:181]: I didnt try to optimize the code yet.

File: Objects/longobject.c (hunk 3 above)
Comment (vstinner) [reply Objects/longobject.c:31]: There is likely room for optimization 😁 I used unbox since it was the simplest fix.

**Comment by vstinner** [comment]
I completed the documentation: https://github.com/vstinner/cpython/blob/tagged_ptr/TAGGED_POINTERS.rst

Benchmark results are not exciting: this PR makes CPython around 1.1x slower on 3 benchmarks.

Results at commit d7348e0aad "longobject.c get_tagged_value()":

```
+-----------+---------+------------------------------+
| Benchmark | ref     | tagptr                       |
+===========+=========+==============================+
| deltablue | 10.8 ms | 12.3 ms: 1.13x slower (+13%) |
+-----------+---------+------------------------------+
| go        | 364 ms  | 400 ms: 1.10x slower (+10%)  |
+-----------+---------+------------------------------+
| telco     | 8.49 ms | 9.44 ms: 1.11x slower (+11%) |
+-----------+---------+------------------------------+
```

**Comment by vstinner** [comment]
I rebased my PR.

**Pull request closed without merging**